File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/generating/MinSpanningTree.java
package com.joeshuff.dddungeongenerator.generator.generating;

import android.graphics.Point;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

public class MinSpanningTree {

    private static boolean longCorridors = false;

    public static void setLongCorridors(boolean longCorridors) {
        MinSpanningTree.longCorridors = longCorridors;
    }

    public static List<Edge> calculate(List<Edge> edges, List<Point> points) {
        List<Edge> mstree = new ArrayList<>();

        if (longCorridors) {
            Collections.sort(edges, Collections.reverseOrder());
        } else {
            Collections.sort(edges);
        }

        int amountofEdges = edges.size();
        while (amountofEdges > 0) {
            Edge minimum = edges.get(0);
            edges.remove(0);

            if (!addingCausesLoop(points, mstree, minimum)) {
                mstree.add(minimum);
            }

            amountofEdges --;
        }

        return mstree;
    }

    // A recursive function that uses visited[] and parent to detect
    // cycle in subgraph reachable from vertex v.
    private static boolean isCyclicUtil(int v, Boolean visited[], int parent, List<Integer>[] adjacent)
    {
        // Mark the current node as visited
        visited[v] = true;
        int i;

        // Recur for all the vertices adjacent to this vertex
        Iterator<Integer> it = adjacent[v].iterator();
        while (it.hasNext())
        {
            i = it.next();

            // If an adjacent is not visited, then recur for that
            // adjacent
            if (!visited[i])
            {
                if (isCyclicUtil(i, visited, v, adjacent))
                    return true;
            }

            // If an adjacent is visited and not parent of current
            // vertex, then there is a cycle.
            else if (i != parent)
                return true;
        }
        return false;
    }

    // Returns true if the graph contains a cycle, else false.
    private static boolean isCyclic(List<Integer>[] adjacent)
    {
        // Mark all the vertices as not visited and not part of
        // recursion stack
        Boolean visited[] = new Boolean[adjacent.length];
        for (int i = 0; i < adjacent.length; i++)
            visited[i] = false;

        // Call the recursive helper function to detect cycle in
        // different DFS trees
        for (int u = 0; u < adjacent.length; u++)
            if (!visited[u]) // Don't recur for u if already visited
                if (isCyclicUtil(u, visited, -1, adjacent))
                    return true;

        return false;
    }

    private static boolean addingCausesLoop(List<Point> points, List<Edge> tree, Edge newEdge) {
        List<Edge> tempTree = new ArrayList<>();
        tempTree.addAll(tree);

        tempTree.add(newEdge);

        List<Integer>[] adjacent = new List[points.size()];

        for (int i = 0; i < points.size(); i ++) {
            adjacent[i] = new ArrayList<>();
        }

        for (Edge e : tempTree) {
            int index = 0;

            for (Point p : points) {

                if (p == e.start && !adjacent[index].contains(e.end)) adjacent[index].add(points.indexOf(e.end));
                if (p == e.end && !adjacent[index].contains(e.start)) adjacent[index].add(points.indexOf(e.start));

                index ++;
            }
        }

        return isCyclic(adjacent);
    }

    public static class Edge implements Comparable<Edge> {
        private Point start;
        private Point end;
        private double distance;

        public Edge(Point start, Point end) {
            this.start = start;
            this.end = end;
            calculateDistance();
        }

        public Edge translate(Point p) {
            return new Edge(
                    new Point(this.start.x + p.x, this.start.y + p.y),
                    new Point(this.end.x + p.x, this.end.y + p.y)
            );
        }

        public void calculateDistance() {
            distance = Math.sqrt(Math.pow(start.x - end.x, 2) + Math.pow(start.y - end.y, 2 ));
        }

        public double getDistance() {
            return distance;
        }

        public Point getStart() {
            return start;
        }

        public Point getEnd() {
            return end;
        }

        @Override
        public int compareTo(Edge o) {
            return (int) (getDistance() - o.getDistance());
        }

        @Override
        public boolean equals(Object obj) {
            if (!(obj instanceof Edge))return false;
            Edge other = (Edge) obj;
            return  ((start == other.start || start == other.end) && (end == other.start || end == other.end));
        }
    }
}

File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/generating/MinSpanningTree.kt
package com.joeshuff.dddungeongenerator.generator.generating

import android.graphics.Point
import java.util.*
import kotlin.math.pow
import kotlin.math.sqrt

object MinSpanningTree {
    private var longCorridors = false

    fun setLongCorridors(longCorridors: Boolean) {
        MinSpanningTree.longCorridors = longCorridors
    }

    fun calculate(edges: ArrayList<Edge>, points: List<Point>): ArrayList<Edge> {
        val mstree: ArrayList<Edge> = ArrayList()

        edges.sort()

        if (longCorridors) {
            edges.reverse()
        }

        var amountofEdges = edges.size
        while (amountofEdges > 0) {
            val minimum = edges[0]
            edges.removeAt(0)

            if (!addingCausesLoop(points, mstree, minimum)) {
                mstree.add(minimum)
            }

            amountofEdges--
        }

        return mstree
    }

    // A recursive function that uses visited[] and parent to detect
    // cycle in subgraph reachable from vertex v.
    private fun isCyclicUtil(v: Int, visited: BooleanArray, parent: Int, adjacent: Array<ArrayList<Int>>): Boolean {
        // Mark the current node as visited
        visited[v] = true
        var i: Int

        // Recur for all the vertices adjacent to this vertex
        val it: Iterator<Int> = adjacent[v].iterator()
        while (it.hasNext()) {
            i = it.next()

            // If an adjacent is not visited, then recur for that
            // adjacent
            if (!visited[i]) {
                if (isCyclicUtil(i, visited, v, adjacent)) return true
            } else if (i != parent) return true
        }
        return false
    }

    // Returns true if the graph contains a cycle, else false.
    private fun isCyclic(adjacent: Array<ArrayList<Int>>): Boolean {
        // Mark all the vertices as not visited and not part of
        // recursion stack
        val visited = BooleanArray(adjacent.size) {false}

        // Call the recursive helper function to detect cycle in
        // different DFS trees
        for (u in adjacent.indices) if (!visited[u]) // Don't recur for u if already visited
            if (isCyclicUtil(u, visited, -1, adjacent)) return true
        return false
    }

    @Suppress("TYPE_INFERENCE_ONLY_INPUT_TYPES_WARNING")
    private fun addingCausesLoop(points: List<Point>, tree: List<Edge>, newEdge: Edge): Boolean {
        val tempTree: MutableList<Edge> = ArrayList()
        tempTree.addAll(tree)

        tempTree.add(newEdge)

        val adjacent: Array<ArrayList<Int>> = Array(points.size) {ArrayList<Int>()}

        for (e in tempTree) {
            points.forEachIndexed { index, point ->
                if (point === e.start && !(adjacent[index] as ArrayList<Int>).contains(e.end)) adjacent[index].add(points.indexOf(e.end))
                if (point === e.end && !(adjacent[index] as ArrayList<Int>).contains(e.start)) adjacent[index].add(points.indexOf(e.start))
            }
        }

        return isCyclic(adjacent)
    }

    class Edge(val start: Point, val end: Point) : Comparable<Edge> {
        var distance = 0.0
            private set

        fun translate(p: Point): Edge {
            return Edge(
                    Point(start.x + p.x, start.y + p.y),
                    Point(end.x + p.x, end.y + p.y)
            )
        }

        fun calculateDistance() {
            distance = sqrt((start.x - end.x.toDouble()).pow(2.0) + (start.y - end.y.toDouble()).pow(2.0))
        }

        override fun compareTo(o: Edge): Int {
            return distance.compareTo(o.distance)
        }

        override fun equals(obj: Any?): Boolean {
            if (obj !is Edge) return false
            return (start === obj.start || start === obj.end) && (end === obj.start || end === obj.end)
        }

        init {
            calculateDistance()
        }
    }
}